### ISSUE TITLE
Memory usage with large regex-based blocklists

### ISSUE THREAD

**Issue by priitohlo** [0xERR0R/blocky#1222]
I'm using Blocky in a container on a Mikrotik RB5009 which is an ARM-based router with 1 GB of memory.
Generally, with plain domain lists (even large ones) it works great but with large regex based lists (e.g. this one: https://big.oisd.nl/regex) I'm seeing that the memory usage of the container starts growing really fast until the device is out of memory.

Is this behaviour with regex-based lists expected and normal or is it fixable in Blocky's code? Am happy to test if need be.

**Comment by kwitsch** [comment]
I looked into it and it seems that the overhead storing the regex as structs accumulate quite a bit.  

There are multiple options to reduce it:
- storing the regex as string and compile them during matching(reduces memory usage / slows matching down)
- fusing all regex to one big(increases load and memory usage during refresh / speed up matching)
- use a different regex engine(most external ones match faster / memory usage has to be evaluated)

**Comment by 0xERR0R** [comment]
Using such big regex files (over 200K in this case) has very poor performance: we match each regex in a loop that means, for each request we have to do 200K regex evaluations

**Comment by priitohlo** [comment]
Maybe support for wildcard domains would be an option here? OISD also provides that list in a wildcard format, which Blocky, sadly, does not support.

**Comment by kwitsch** [comment]
> I also thought a bit about this and came up with the same ideas as @kwitsch.
> 
> > - storing the regex as string and compile them during matching(reduces memory usage / slows matching down)
> 
> I think we can eliminate this option from the start, or at least try everything else first, cause that sounds very slow.

Agreed 👍 
I only included it for completeness.

> > - fusing all regex to one big(increases load and memory usage during refresh / reduce long term memory usage)
> > - use a different regex engine(most external ones match faster / memory usage has to be evaluated)
> 
> I think my preferred option would be to combine both of these. We could start with just the first one, and then switch the engine we use to improve performance and hopefully further improve memory usage.  

Thought the same. 👍

> To build the pattern, we should definitely use a `string.Builder` to avoid growing a buffer in a loop and only allocating the giant sting at the end.

Agreed 👍

> Ideally the engine would have an API that lets us assemble the regex so we can avoid handling a `string.Builder` and regex syntax ourselves. The idea being such an API could be optimized to never actually build the full string since the regex engine compiles it to a state machine anyways. The exception being to print the pattern, but seems like a fair tradeoff.  
> Don't think I've ever seen such an API though, so I don't have too much hope for avoid a `string.Builder`.

This seems to be an optimization for later on so let's start with strings. 😉

> Here's a quick look at various regex engines I found:
> - https://github.com/AspieSoft/go-regex adds a cache -> will use more memory
> - https://github.com/AspieSoft/go-regex-re2 same as above
> - https://github.com/admpub/regexp2 says "You'll likely be better off with the RE2 engine from the regexp package"
> - https://github.com/GRbit/go-pcre worth a try. Provides JIT and no JIT options, would be interested in seeing memory usage difference between those

This one performed very well in the benchmarks I looked at: https://github.com/BurntSushi/rure-go

> Another completely different path forward I haven't investigated: write a very basic custom regex engine with the above API. Lists that have regex most likely don't need any fancy regex features (needs to be confirmed) so we could probably get away with something simple.

I strongly advise against that.
Did something similar in the past and switch back later on since it got quite complex really fast.

**Comment by ThinkChaos** [comment]
I did a small test of the giant regex idea and it reduces memory usage by >25%: when using only the OISD big regex list, at startup from 436 MB to 303 MB.  Not sure how that difference scales cause I only tried that one list, but seems worth it.  
Made PR #1227 with that code.

The downside is we can't easily log which regex matched anymore.  
See my comment in #558 on a potential solution.

> This one performed very well in the benchmarks I looked at: https://github.com/BurntSushi/rure-go

It says the overhead of C FFI is not worth it for small haystacks, which applies to us. Though maybe having a giant regex would flip that back to favorable.  
I guess the same would apply to [go-pcre](https://github.com/GRbit/go-pcre).

> I strongly advise against that. Did something similar in the past and switch back later on since it got quite complex really fast.

From a quick look at the OISD regex list, everything is in the format `/^(.*\.)?0\-\-foodwarez\.da\.ru$/` which is trivially rewritable to a wildcard.

The more I think about all this, the more I'm convinced we should remove regex support for non inline lists, or at least warn about memory usage in docs and not spend more time on the problem.  
Case in point: OISD formatted as domains uses 16 MB RAM, compared to the 436/303 MB mentioned above! A wildcard solution would be much closer to 16 MB than 300.

@priitohlo I know we don't have support for any non deprecated OISD format besides regex but my recommendation is to use a deprecated one for now.
